update clasp

BenKaufmann opened the pull request:
* add colored help output and fix regression in option parser
* extend text output

@rkaminsk  This update is also a first step towards custom text output. 
For example, the output of `clingcon` could now be achieved by:

- Adding options  `--out-assign=__csp/2` and `--out-cost=__csp_cost/1:%0`
- Adjusting `Clingcon::Propagator::on_model() ` so that it adds the cost as a `Clingo::Number(lib_, std::to_string(bound));` instead of a `Clingo::String`.

For `clingo-lpx`, it would be:
- `--out-assign=__lpx/2` and `--out-cost=__lpx_objective/2:%0 [%1]`
- Adjusting `LPXPropagatorFacade::extend_model()`  so that it adds `Clingo::Number`s instead of strings. 

**TODO/RFC**

- There is an API missing for properly setting output options like `--out-assign`.
- The theory assignment extracted from the theory atoms is not yet sorted.
- Atom arguments are always printed verbatim (including quotes). E.g. for a theory atom `__lpx_objective(-132, "bounded")` and `--out-cost=__lpx_objective/2:%0 [%1]`, `clasp` currently prints `Costs: -132 ["bounded"]`. Hence, maybe it should be possible to extend the argument references with a type/print-hint (e.g. `%1s`).

BenKaufmann commented:
> **TODO/RFC**
> 
>     * There is an API missing for properly setting output options like `--out-assign`.
>     * The theory assignment extracted from the theory atoms is not yet sorted.
>     * Atom arguments are always printed verbatim (including quotes). E.g. for a theory atom `__lpx_objective(-132, "bounded")` and `--out-cost=__lpx_objective/2:%0 [%1]`, `clasp` currently prints `Costs: -132 ["bounded"]`. Hence, maybe it should be possible to extend the argument references with a type/print-hint (e.g. `%1s`).

Regarding the last two points: Maybe it is enough to
1. sort (theory) symbols in `ModelExtend::first()`
2. always _unquote_ arguments in `--out-assign`/`--out-cost`. One could then use `"%0"` to force quoted output if needed.

rkaminsk commented:
1. With theory atoms, we could add atoms already in sorted order.
2.  It would be nice to omit quotes of strings in output. For clingo lpx, the objective and values of the assignment cannot always be represented as a 32bit integer. The system works with arbitrary precision rationals. So there is probably no way around a string.
3. Would it be possible to add the options out-assign to the existing configuration object, e.g., register some app options with it?

BenKaufmann committed:
update clasp

* add colored help output and fix regression in option parser
* extend text output

BenKaufmann committed:
let ModelExtend return symbols in sorted order

BenKaufmann committed:
add API for setting an option's default value

* add `clingo_options_set_default_value` to app C-API and adjust
  C++-API accordingly

* simplify AppOptions

BenKaufmann commented:
@rkaminsk FYI: I adjusted the PR so that it now addresses the three mentioned issues.
[This commit](https://github.com/potassco/clingcon/commit/9bbd460cf9115d2bd02c4009cad69c2527308567)  shows how one could use the new API / TextOutput in clingcon.

BenKaufmann committed:
add `set_default_value` to python api

rkaminsk commented:
This actually looks quite nice. I'll try it out in the other systems.

rkaminsk commented on lib/python-api/tests/test_app.py:
```diff
@@ -131,17 +139,25 @@ def test_app(self):
-            ),
-            ("models", [["a"], ["a", "b"], ["b"]]),
-        ]
+        with NamedTemporaryFile(mode="wt", delete=False) as fp:
+            name = fp.name
```
And then get warnings that mktemp is unsafe? Better not!

rkaminsk committed:
add tmode example for testing extended output config

rkaminsk commented:
In d60a350, I added the tmode example from the old clingo, which has similar output as compared to telingo system. We could try to see how to adjust the output to nicely output it's models.

BenKaufmann commented:
> In [d60a350](https://github.com/potassco/clingo/commit/d60a3507bea8c0c6a4a828554df56d9b9f4e9f54), I added the tmode example from the old clingo, which has similar output as compared to telingo system. We could try to see how to adjust the output to nicely output it's models.

Currently, the model output for `python tmode.py example.lp` is:

```
Answer: 1 (Time: 0.009s)
 State 0:
  q(0)
  r(0)
 State 1:
  q(1)
  r(0) r(1)
 State 2:
  q(2)
  r(0) r(1) r(2)
 State 3:
  q(3)
  r(0) r(1) r(2) r(3)
 State 4:
  q(4)
  r(0) r(1) r(2) r(3) r(4)
SATISFIABLE
```
With the clasp changes in this PR and adding 
```python
def register_options(self, options: AppOptions) -> None:
        options.set_default_value("out-pred-sep", "\n")
        options.set_default_value("out-step", "last")
``` 
to the TModeApp, you'll get nearly identical output without having to write a `print_model()` function.

<img width="500" height="270" alt="new" src="https://github.com/user-attachments/assets/15e9f36e-4895-4f41-a22b-854b8fe494ce" />

Two things are missing, though:
1. There is currently no way to set the "Step" string - the original TModeApp uses "State" instead of "Step"
2. There is currently no way to filter out certain atoms (neither by name nor by arity) - the original TModeApp does not print "query"

Any suggestions how to best handle the second point?

**FYI:** example2 fails for me with a `RuntimeError: RuntimeError: not implemented`.

rkaminsk commented:
> There is currently no way to set the "Step" string - the original TModeApp uses "State" instead of "Step"

It would be nice to be able to configure this.

> Any suggestions how to best handle the second point?

I am not sure what the best way here. This would need some way to hide predicates. The question is what is the best place to implement this. Add a generic way to hide atoms with clingo or maybe add a filter option to clasp's text-output: something like `--out-hide=query/1`.

> FYI: example2 fails for me with a RuntimeError: RuntimeError: not implemented.

Yes, a known issue and I was already working on it but got interrupted. Just something that still has to be implemented.

EDIT: the example should be in a nice shape now. All that is left is to get rid of the auxiliary atoms in the output.

rkaminsk committed:
make the query atom of the incmode configurable

This commit adds option `--iquery=<arg>` to configure the name of the
query atom. This is used in the tmode example to use the name finally in
the example.

rkaminsk commented:
I added an option to change the name of the query atom. Now, I actually don't find the output including initially and finally so bad:
<img width="807" height="672" alt="image" src="https://github.com/user-attachments/assets/c3e614c7-9e7b-4cc9-abdd-d3995e6c0740" />

I am still thinking to add a `#hide` again because this use case shows that there are certain use cases that can actually benefit from such a statement.

rkaminsk committed:
add a flag to show statements to show/hide predicates

rkaminsk commented:
With the latest commit we are here:

<img width="834" height="628" alt="image" src="https://github.com/user-attachments/assets/0de5eae4-830e-4dcb-8060-c9188a5fa0ca" />

BenKaufmann committed:
update clasp and address todo in tmode example

BenKaufmann commented:
With the latest commit, `tmode` now prints states:
<img width="500" height="405" alt="tmode" src="https://github.com/user-attachments/assets/3b8e9f60-f512-4b9e-a841-8ccdd11f5463" />

rkaminsk commented:
I really like the current state. We can merge if you are okay work it.
